Update FVA to include new algorithms for implementing the loop law

shjchan opened the pull request:
*Please include a short description of enhancement here*
This is a huge update for `fluxVariability.m` regarding the implementation of the loop law:
- Minimize the call of `addLoopLawConstraints.m` because it is time-consuming to build the MILP problem every time
- Implement three newer versions of loop law, available through the flag `allowLoops`:
  Fast-SNP (Saa and Nielsen, 2016), localized loopless constraints using null-space or elementary flux modes (Chan et al., 2018)
- Implement `minNorm` (2-norm, 1-norm and 0-norm) along with loop law
- Update `addLoopLawConstraints.m` to support the new implementation
- Add `fastSNP.m` for finding a minimal nullspace using the Fast-SNP method
- Add `findMinNull.m` for finding a minimal nullspace quickly by solving MILPs

References:
Saa, P. A., & Nielsen, L. K. (2016). Fast-SNP: a fast matrix pre-processing algorithm for efficient loopless flux optimization of metabolic models. Bioinformatics, 32(24), 3807-3814.
Chan, S. H., Wang, L., Dash, S., & Maranas, C. D. (2018). Accelerating flux balance calculations in genome-scale metabolic models by localizing the application of loopless constraints. Bioinformatics, 34(24), 4248-4255.

**I hereby confirm that I have:**

- [X] Tested my code on my own machine
- [X] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [X] Selected `develop` as a target branch (top left drop-down menu)

*(Note: You may replace [ ] with [X] to check the box)*

rmtfleming commented:
Minimal nullspace in what sense?
How does it compare with minSpan?
https://opencobra.github.io/cobratoolbox/latest/tutorials/tutorialMinSpan.html


On Fri, 15 Mar 2019 at 21:53, Joshua Chan <notifications@github.com> wrote:

> *Please include a short description of enhancement here*
> This is a huge update for fluxVariability.m regarding the implementation
> of the loop law:
>
>    - Minimize the call of addLoopLawConstraints.m because it is
>    time-consuming to build the MILP problem every time
>    - Implement three newer versions of loop law, available through the
>    flag allowLoops:
>    Fast-SNP (Saa and Nielsen, 2016), localized loopless constraints using
>    null-space or elementary flux modes (Chan et al., 2018)
>    - Implement minNorm (2-norm, 1-norm and 0-norm) along with loop law
>    - Update addLoopLawConstraints.m to support the new implementation
>    - Add fastSNP.m for finding a minimal nullspace using the Fast-SNP
>    method
>    - Add findMinNull.m for finding a minimal nullspace quickly by solving
>    MILPs
>
> References:
> Saa, P. A., & Nielsen, L. K. (2016). Fast-SNP: a fast matrix
> pre-processing algorithm for efficient loopless flux optimization of
> metabolic models. Bioinformatics, 32(24), 3807-3814.
> Chan, S. H., Wang, L., Dash, S., & Maranas, C. D. (2018). Accelerating
> flux balance calculations in genome-scale metabolic models by localizing
> the application of loopless constraints. Bioinformatics, 34(24), 4248-4255.
>
> *I hereby confirm that I have:*
>
>    - Tested my code on my own machine
>    - Followed the guidelines in the Contributing Guide
>    <https://opencobra.github.io/cobratoolbox/docs/contributing.html>
>    - Selected develop as a target branch (top left drop-down menu)
>
> *(Note: You may replace [ ] with [X] to check the box)*
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/opencobra/cobratoolbox/pull/1446
> Commit Summary
>
>    - Add options to different methods for loopless FVA
>    - Update addLoopLawConstraints accordingly
>    - Functions for quickly finding minimall nullspace
>    - Fix parallel part
>    - Test both single-thread and parallel
>    - To ensure the exact same problem structure as before for the
>    original method
>    - Support minNorm (2-, 1-, 0-norms) with allowLoops on
>    - Update testFVA to test both single-thread and parallel computation
>    - Add test for loopless FVA
>    - Update comments
>    - Update test (unfinished)
>    - Update docs, comments and fix bugs
>    - Update comments in addLoopLawConstraints.m
>    - Use sparseNull directly in findMinNull.m for that to work
>    - Add test for looplessFVA
>
> File Changes
>
>    - *M* src/analysis/FVA/fluxVariability.m
>    <https://github.com/opencobra/cobratoolbox/pull/1446/files#diff-0>
>    (673)
>    - *M* src/analysis/thermo/thermoFBA/addLoopLawConstraints.m
>    <https://github.com/opencobra/cobratoolbox/pull/1446/files#diff-1>
>    (207)
>    - *A* src/analysis/thermo/thermoFBA/fastSNP.m
>    <https://github.com/opencobra/cobratoolbox/pull/1446/files#diff-2>
>    (135)
>    - *A* src/analysis/thermo/thermoFBA/findMinNull.m
>    <https://github.com/opencobra/cobratoolbox/pull/1446/files#diff-3>
>    (389)
>    - *A* test/verifiedTests/analysis/testFVA/refData_looplessFVA.mat
>    <https://github.com/opencobra/cobratoolbox/pull/1446/files#diff-4> (0)
>    - *M* test/verifiedTests/analysis/testFVA/testFVA.m
>    <https://github.com/opencobra/cobratoolbox/pull/1446/files#diff-5>
>    (369)
>    - *A* test/verifiedTests/analysis/testFVA/testFVAloopless.m
>    <https://github.com/opencobra/cobratoolbox/pull/1446/files#diff-6>
>    (180)
>
> Patch Links:
>
>    - https://github.com/opencobra/cobratoolbox/pull/1446.patch
>    - https://github.com/opencobra/cobratoolbox/pull/1446.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobratoolbox/pull/1446>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACDCOkeMNi6SjpuEuk0azkAXhZFJ0kLrks5vXBZCgaJpZM4b3XP7>
> .
>


-- 
--
Mr. Ronan MT Fleming B.V.M.S. Dip. Math. Ph.D.
----------------------------------------------------------------------------
Assistant Professor,
Division of Systems Biomedicine and Pharmacology,
Leiden Academic Centre for Drug Research,
Faculty of Science,
Leiden University.
https://www.universiteitleiden.nl/en/staffmembers/ronan-fleming
&
H2020 Project Coordinator
Systems Medicine of Mitochondrial Parkinson’s Disease
http://sysmedpd.eu
----------------------------------------------------------------------------
Peer-reviewed publications: https://goo.gl/FZPG23
Mobile:  +353 873 413 072
Skype: ronan.fleming
----------------------------------------------------------------------------
(This message is confidential and may contain privileged information. It is
intended for the named recipient only. If you receive it in error please
notify me and permanently delete the original message and any copies.)

shjchan commented:
@rmtfleming , thanks for referring me to that. From a brief look, `detMinSpan`  seems almost exactly the same as `fastSNP` implemented in Saa and Nielsen (2016). But `fastSNP` implemented here is only meant for finding the internal null space, i.e., the basis for all the loops. However, `findMinNull` is a little bit different, though trying to find the same thing as `fastSNP`. It solves some LPs and then an MILP to determine the minimal null space (smallest null space that contains the entire feasible flux space). It has an advantage regarding scalability. As the dimension of the null space grows, Fast-SNP (and probably minSpan though I am not 100% sure) takes a long time to loop and find one basis vector per iteration, while `findMinNull` is much faster than `fastSNP` for very large models.

shjchan commented:
Some more data about the performance:
For the 50+ reactions in E. coli iJR904 in `testFVAloopless`, the time taken:
Previous implementation of the original loop law: 50 s
Updated implementation of the original loop law: 16 s
Fast-SNP: 4.2 s
LLC-NS: 2.6 s
LLC-EFM: 3 s

This COBRA script was also tested for the E. coli iJO1366 model.
Previous implementation of the original loop law: not tested
Updated implementation of the original loop law: 650 s
Fast-SNP: 53 s
LLC-NS: 24 s
LLC-EFM: 25 s

The FVA results of all methods tested have insignificant differences (<1e-6). Basically the new algorithms outperform the old implementation by ~10 - 20 x

shjchan commented:
@laurentheirendt , I have actually further worked on `fluxVariability.m` to make it support parameter-value input and introduced a function `parseCobraVarargin` that I believe to be general enough for parsing `varargin` when writing cobra functions (I think no similar function exists yet?) Care to [take a look](https://github.com/shjchan/cobratoolbox/compare/update_looplessFVA...shjchan:name_value_input_FVA)? The function is modified from what is in `computeMetFormulae` that @tpfau and I have worked on. What is the best way for me to proceed? Another pull request after this one? Or directly merging to this branch will be okay?

rmtfleming commented:
Hi Joshua,
nice work. Please ensure that the code is still backward compatible. Not
everyone likes working in the style of param:value pairs, esp for long
sequences of conjoined function calls, so it is good to have optional usage
approaches that are tested to all work the same.
Regards,
Ronan

On Sat, 23 Mar 2019 at 17:00, Joshua Chan <notifications@github.com> wrote:

> @laurentheirendt <https://github.com/laurentheirendt> , I have actually
> further worked on fluxVariability.m to make it support parameter-value
> input and introduced a function parseCobraVarargin that I believe to be
> general enough for parsing varargin when writing cobra functions (I think
> no similar function exists yet?) Care to take a look
> <https://github.com/shjchan/cobratoolbox/compare/update_looplessFVA...shjchan:name_value_input_FVA>?
> The function is modified from what is in computeMetFormulae that @tpfau
> <https://github.com/tpfau> and I have worked on. What is the best way for
> me to proceed? Another pull request after this one? Or directly merging to
> this branch will be okay?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobratoolbox/pull/1446#issuecomment-475886532>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACDCOne3vJ5vFC0xdxs5WVIkkLcYwQcfks5vZl3IgaJpZM4b3XP7>
> .
>


-- 
--
Mr. Ronan MT Fleming B.V.M.S. Dip. Math. Ph.D.
----------------------------------------------------------------------------
Assistant Professor,
Division of Systems Biomedicine and Pharmacology,
Leiden Academic Centre for Drug Research,
Faculty of Science,
Leiden University.
https://www.universiteitleiden.nl/en/staffmembers/ronan-fleming
&
H2020 Project Coordinator
Systems Medicine of Mitochondrial Parkinson’s Disease
http://sysmedpd.eu
----------------------------------------------------------------------------
Peer-reviewed publications: https://goo.gl/FZPG23
Mobile:  +353 873 413 072
Skype: ronan.fleming
----------------------------------------------------------------------------
(This message is confidential and may contain privileged information. It is
intended for the named recipient only. If you receive it in error please
notify me and permanently delete the original message and any copies.)

shjchan commented:
Hi Ronan, 

Yes, sure. I always have that in mind. And I actually like using mixed way of inputs, just for convenience. The modifications are intended for 100% backward compatibility. 

Best, 
Joshua

laurentheirendt commented on src/analysis/thermo/thermoFBA/addLoopLawConstraints.m:
```diff
@@ -259,7 +332,77 @@
+end
 end
+
+function rxnLink = getRxnLink(model, conComp, rxnInLoops)
```
May you split out these functions into separate ones?

rmtfleming replied:
todo

laurentheirendt commented on src/analysis/thermo/thermoFBA/fastSNP.m:
```diff
@@ -0,0 +1,135 @@
+function [N, resInfo] = fastSNPcobra(model, varargin)
```
This function does actually not correspond to the name of the file.

rmtfleming replied:
I will fix this separately.

shjchan commented:
Hi, I think most are fixed now. Have also changed the test to use a toy model with sufficient features so that the test run time will not be horribly long. Let me know if changes/clarifications are needed anywhere in the code.

rmtfleming commented:
I attempted a manual resolution of the conflicts taking shjchan/update_looplessFVA in preference to develop.
